src/Common/Auth/OpenIDConnect/Grant/CustomClientCredentialsGrant.php

```diff
@@ -0,0 +1,369 @@
+
+    /**
+     * The required value for the jwt assertion type
+     */
+    const OAUTH_JWT_CLIENT_ASSERTION_TYPE = 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer';
+
+    /**
+     * CustomClientCredentialsGrant constructor.
+     * @param $authTokenUrl string The OAUTH2 token issuing url to be used as the audience parameter for JWT validation
+     */
+    public function __construct($authTokenUrl)
+    {
+        $this->logger = new SystemLogger(); // default if we don't have one.
+        $this->authTokenUrl = $authTokenUrl;
+        $this->trustedUserService = new TrustedUserService();
+        $this->userService = new UserService();
+    }
+
+    /**
+     * @return LoggerInterface
+     */
+    public function getLogger(): LoggerInterface
+    {
+        return $this->logger;
+    }
+
+    /**
+     * @param LoggerInterface $logger
+     */
+    public function setLogger(LoggerInterface $logger): void
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * Allows the http client that retrieves jwks to be overriden.  Useful for unit testing
+     * @param ClientInterface $client
+     */
+    public function setHttpClient(ClientInterface $client)
+    {
+        $this->httpClient = $client;
+    }
+
+    /**
+     * Allows the http client that retrieves jwks to be overriden.  Useful for unit testing
+     * @return ClientInterface
+     */
+    public function getHttpClient(): ClientInterface
+    {
+        return $this->httpClient;
+    }
+
+    /**
+     * @return UserService
+     */
+    public function getUserService(): UserService
+    {
+        return $this->userService;
+    }
+
+    /**
+     * @param UserService $userService
+     */
+    public function setUserService(UserService $userService): void
+    {
+        $this->userService = $userService;
+    }
+
+    /**
+     * We issue an access token, but we force the user account to be our OpenEMR API system user.  We also save off the
+     * grant as a TrustedUser which we can use later for revocation if necessary.
+     * @param DateInterval $accessTokenTTL
+     * @param ClientEntityInterface $client
+     * @param string|null $userIdentifier
+     * @param array $scopes
+     * @return \League\OAuth2\Server\Entities\AccessTokenEntityInterface
+     * @throws OAuthServerException
+     * @throws \League\OAuth2\Server\Exception\UniqueTokenIdentifierConstraintViolationException
+     */
+    protected function issueAccessToken(DateInterval $accessTokenTTL, ClientEntityInterface $client, $userIdentifier, array $scopes = [])
+    {
+        // let's grab our user id here.
+        if ($userIdentifier === null) {
+            // we want to grab our system user
+            $systemUser = $this->userService->getSystemUser();
+            $userIdentifier = $systemUser['uuid'] ?? null;
+        }
+        $accessToken = parent::issueAccessToken($accessTokenTTL, $client, $userIdentifier, $scopes); // TODO: Change the autogenerated stub
```

Comment:
Are we ensuring that only those 4 system scopes (i think there were 4) are allowed in this grant request (ie. not allow other system scopes or patient/user scopes)?

Reply:
I've pushed a fix out that turns off all system scopes unless someone enables the system scope flag.  People can actually request system scopes using Authorization Code grant, Password grant, or Credentials grant.  The systems scope flag handles all those cases.